run stop_crio inside cleanup_test

edsantiago opened the pull request:
If a subtest fails it might leave crio running; this prevents
bats from exiting, causing a hang. It also prevents cleanup of
the test directory because there's an active mount in it.

Solution: run stop_crio from within cleanup_test, just before
the rm -rf. This is safe because stop_crio checks for an active
crio pid.

Signed-off-by: Ed Santiago <santiago@redhat.com>

edsantiago commented:
Note: although this is short & sweet, I can't help but think that there's a reason this hasn't already been done...

rhatdan commented:
LGTM
@mrunalp @runcom Any idea why this was not done before?

wking commented on test/helpers.bash:
```diff
@@ -318,6 +318,7 @@ function restart_crio() {
 }
 
 function cleanup_test() {
+	stop_crio
```
If you add it here you can remove it from all the other places where it's currently called before `cleanup_test`, e.g. [here][1], with `cleanup_test` [here][2].  That may turn up a reason against this (e.g. a test that gets `cleanup_test` but no `start_crio`).

[1]: https://github.com/kubernetes-incubator/cri-o/blob/959afb79d4247891b6176b74609b1133cb939f1b/test/apparmor.bats#L36
[2]: https://github.com/kubernetes-incubator/cri-o/blob/959afb79d4247891b6176b74609b1133cb939f1b/test/apparmor.bats#L6

edsantiago replied:
Oh... interesting. Your first example shows what I think might be the crux of this issue: those cleanups at the end of a subtest will never get run if an intermediate step fails. My hunch is that all those `cleanup_xx` and `remove_apparmor_profile` steps should be moved into `teardown()`, possibly with safeguards added so they pass cleanly if unneeded (I haven't looked). Thanks for pointing this out.

edsantiago commented:
Rebased, force-pushed, and added a fixup commit addressing the issue reported by @wking. The main addition is running `cleanup_ctrs` and `cleanup_pods` as part of final cleanup.

I've tested this on a successful test run and an unsuccessful one (bad `crictl`). In both cases I see no new errors; in the latter, I see improved behavior: first, the test terminates (the original intention behind this PR), and second, the following error is no longer seen on test failure:

    rm: cannot remove '/tmp/tmp.lhsW8REnI8/crio/devicemapper/mnt/f849223dde67d1d0fcb5ddba526b520de079055691c296cef242f1c365f12c8f': Device or resource busy

(And, as then expected, the test system is not left in a state with dozens of /tmp/tmp.*/crio/devicemapper mounts).

Since this is a small change with potential ramifications I haven't considered, I'd like some eyeballs and perhaps test runs on this. Should this look OK to all, I'd welcome opinions on whether to remove `cleanup_xxxx` and `stop_crio` from the various teardowns, or leave them. My (slight) preference is to remove them, because they're unnecessary and because, honestly, they don't seem to be working: some modules seem to be missing required cleanup calls.

Thank you.

rhatdan commented:
Does cleanup_test get called after every test or just once at the end of the run?  My concern with removing the cleanup_xxx calls in the other tests would be that they end up dirtying up other tests.

edsantiago commented:
Per the BATS docs, `teardown` gets called [after each test case](https://github.com/sstephenson/bats#user-content-setup-and-teardown-pre--and-post-test-hooks). 

>My concern with removing the cleanup_xxx calls in the other tests would be that they end up dirtying up other tests.

My mistake. I was thinking of `network.bats` which has explicit cleanups in its teardown, and thought there were other `.bat` files like that. There aren't: only network.bats has those, and because of the `rm -rf` in it, I think it's probably best to keep those cleanups there. All other `.bat` files have a single-line teardown, `cleanup_test`, which is what I think is correct. So please ignore that part of my comment. I'm sorry.

rhatdan commented:
LGTM

rhatdan commented:
/test all

cevich commented:
Giving this a go manually on 7.5-nightly.

cevich commented:
Update: ``setup`` and ``integration`` passed, running ``node-e2e``...

cevich commented:
Update: ``node-e2e`` test failed (not sure if that's significant, since this is 7.5-nightly):

```
Failure [195.590 seconds]
[BeforeSuite] BeforeSuite 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/e2e_node_suite_test.go:141

  should be able to start node services.
  Expected success, but got an error:
      <*errors.errorString | 0xc420049580>: {
          s: "failed to start kubelet: e2e service \"kubelet\" readiness check timeout 2m0s",
      }
      failed to start kubelet: e2e service "kubelet" readiness check timeout 2m0s

  /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/e2e_node_suite_test.go:162
------------------------------
I0313 16:09:01.147964   26803 e2e_node_suite_test.go:184] Stopping node services...
I0313 16:09:01.147973   26803 server.go:258] Kill server "kubelet"
I0313 16:09:01.148045   26803 server.go:177] Waiting for server "kubelet" start command to complete after initial health check failed
F0313 16:09:01.148073   26803 server.go:181] Restart loop readinessCheck failed for server "kubelet" start-command: `/bin/systemd-run --unit=kubelet-1829939178.service --slice=runtime.slice --remain-after-exit /go/src/k8s.io/kubernetes/_output/local/go/bin/kubelet --kubeconfig /go/src/k8s.io/kubernetes/_output/local/go/bin/kubeconfig --root-dir /var/lib/kubelet --v 4 --logtostderr --allow-privileged true --network-plugin=kubenet --cni-bin-dir /go/src/k8s.io/kubernetes/_output/local/go/bin/cni/bin --cni-conf-dir /go/src/k8s.io/kubernetes/_output/local/go/bin/cni/net.d --hostname-override cevich-crio-testing --container-runtime remote --container-runtime-endpoint /var/run/crio.sock --image-service-endpoint /var/run/crio/crio.sock --config /go/src/k8s.io/kubernetes/_output/local/go/bin/kubelet-config --container-runtime=remote --network-plugin= --cni-bin-dir= --cgroup-driver=systemd`, kill-command: `/bin/systemctl kill kubelet-1829939178.service`, restart-command: `/bin/systemctl restart kubelet-1829939178.service`, health-check: [http://127.0.0.1:10255/healthz], output-file: "kubelet.log"

Ginkgo ran 1 suite in 3m15.888712621s
Test Suite Failed
F0313 16:09:01.158481   25413 run_local.go:70] Test failed: exit status 1
exit status 1
make: *** [test-e2e-node] Error 1
```

regular ``e2e`` tests running now...

rhatdan commented:
@cevich Looks like All 8 tests passed?

cevich commented:
@rhatdan Ed asked for some additional testing of this, so since "all 8 passed", I thought I'd try something different.  Please don't let me hold anything up though.  I wouldn't consider this run authoritative, and I don't know enough that it'll be informative.  *shrug*  Worst case is it stumbles/bumbles into something useful or helpful.  I can always hope :grinning: 

 ``e2e`` is still humming along...I'll check back on it tomorrow.

cevich commented:
Looks like everything except the above ``node-e2e`` test passed.  Attaching results.

[node-e2e.log](https://github.com/kubernetes-incubator/cri-o/files/1811088/node-e2e.log)
[testout.txt](https://github.com/kubernetes-incubator/cri-o/files/1811091/testout.txt)
[junit_01.xml.txt](https://github.com/kubernetes-incubator/cri-o/files/1811099/junit_01.xml.txt)
[e2e.log](https://github.com/kubernetes-incubator/cri-o/files/1811100/e2e.log)

rhatdan commented:
@mrunalp @runcom @wking @umohnani8 PTAL

mrunalp commented:
@edsantiago Can you squash the commits? Right now we have a fixup commit.

edsantiago commented:
@mrunalp done: squashed & rebased

umohnani8 commented:
LGTM

mrunalp commented:
/test all

rhatdan commented:
Passes tests
Merging.
